knowledge-base/kb-pdfviewer-viewport-snapshot.md

```diff
@@ -0,0 +1,169 @@
+
+        public string Name => "ViewportUILayer";
+
+        public void Clear()
+        {
+            MainWindow.VisiblePagesToViewport.Remove(this.page);
+            this.page = null;
+        }
+
+        public void Initialize(UILayerInitializeContext context)
+        {
+            this.page = context.Page;
+        }
+
+        public void Update(UILayerUpdateContext context)
+        {
+            MainWindow.VisiblePagesToViewport[this.page] = context.Viewport;
+        }
+    }
+
+{{endregion}}
+
+The next step is to create a custom [UILayersBuilder](https://docs.telerik.com/devtools/wpf/api/telerik.windows.documents.fixed.ui.layers.uilayersbuilder), which will help us to include the newly created ViewportUILayer into the [IUILayerContainer](https://docs.telerik.com/devtools/wpf/api/telerik.windows.documents.fixed.ui.layers.iuilayercontainer).
+
+#### __[C#] Example 2: Creating the custom UILayersBuilder__
+{{region kb-pdfviewer-viewport-snapshot-1}}
+
+    public class CustomLayersBuilder : UILayersBuilder
+    {
+        protected override void BuildUILayersOverride(IUILayerContainer uiLayerContainer)
+        {
+            base.BuildUILayersOverride(uiLayerContainer);
+
+            uiLayerContainer.UILayers.AddLast(new ViewportUILayer());
+        }
+    }
+
+{{endregion}}
+
+Using the ExtensibilityManager`s [RegisterLayersBuilder()](https://docs.telerik.com/devtools/wpf/api/telerik.windows.documents.fixed.ui.extensibility.extensibilitymanager#collapsible-Telerik_Windows_Documents_Fixed_UI_Extensibility_ExtensibilityManager_RegisterLayersBuilder_Telerik_Windows_Documents_Fixed_UI_Layers_UILayersBuilder_) method we will register the CustomLayersBuilder.
+
+#### __[C#] Example 3: Registering the custom UILayersBuilder__
+{{region kb-pdfviewer-viewport-snapshot-2}}
+
+    public RadFixedDocument document;
+
+    public MainWindow()
+    {
+        this.InitializeComponent();
+        
+        CustomLayersBuilder uILayersBuilder = new CustomLayersBuilder();
+        ExtensibilityManager.RegisterLayersBuilder(uILayersBuilder);
+
+        PdfFormatProvider pdfFormatProvider = new PdfFormatProvider();
+        using (FileStream stream = File.OpenRead("SampleData/SampleDocument.pdf"))
+        {
+            this.pdfViewer.Document = pdfFormatProvider.Import(stream);
+        }
+    }
+
+    static MainWindow()
+    {
+        VisiblePagesToViewport = new Dictionary<RadFixedPage, Rect>();
+    }
+
+    public static Dictionary<RadFixedPage, Rect> VisiblePagesToViewport { get; set; }
+
+    private void CreateSnapshot(object sender, RoutedEventArgs e)
+    {
+        // Implemented in Example 4
+    }
+		
+{{endregion}}
+
+Having the _VisiblePagesToViewport_ already, we can iterate its pages and using the [ThumbnailFactory](https://docs.telerik.com/devtools/wpf/api/telerik.windows.documents.ui.thumbnailfactory) class to export the snapshots in the desired image format (in this example to TIFF images).
```

Comment:
Now when we have the _VisiblePagesToViewport_, we can iterate its pages and export the snapshots to the desired image format using the [ThumbnailFactory](https://docs.telerik.com/devtools/wpf/api/telerik.windows.documents.ui.thumbnailfactory) class. For the purpose of the example, we will generate a TIFF image.